cloudinit/gpg.py

```diff
@@ -68,6 +69,43 @@ def export_armour(self, key: str) -> Optional[str]:
             LOG.debug('Failed to export armoured key "%s": %s', key, error)
         return None
 
+    def import_key(self, key: pathlib.Path) -> None:
+        """Import gpg key from a file to the temporary keyring.
+
+        :param key: path to the key file
+        """
+        try:
+            subp.subp(
+                [
+                    "gpg",
+                    "--batch",
+                    "--import",
+                    str(key),
+                ],
+                update_env=self.env,
+            )
+        except subp.ProcessExecutionError as error:
+            LOG.warning("Failed to import key %s: %s", key, error)
+
+    def decrypt(self, data: str) -> str:
+        """Process data using gpg.
+
+        This can be used to decrypt encrypted data, verify signed data,
```

Comment:
This only verifies signed data if the data is signed, correct?

IOW, if a user wants their user data to be signed and encrypted, an attacker would need only to acquire the public key of the encryption key and then provide their own **unsigned** payload to cloud-init, and cloud-init will not verify it.

Reply:
Thank you. I pushed a new version that checks that gpg verified the signature if we require the signature.